library-core/src/main/java/com/mikepenz/fastadapter/dsl/FastAdapterDsl.kt

```diff
@@ -0,0 +1,31 @@
+package com.mikepenz.fastadapter.dsl
+
+import com.mikepenz.fastadapter.FastAdapter
+import com.mikepenz.fastadapter.GenericItem
+import com.mikepenz.fastadapter.IAdapter
+import com.mikepenz.fastadapter.IItemList
+import com.mikepenz.fastadapter.adapters.ItemAdapter
+import com.mikepenz.fastadapter.adapters.ModelAdapter
+
+@DslMarker
+annotation class FastAdapterDsl
+
+fun genericfastadapter(block: (FastAdapter<GenericItem>).() -> Unit): FastAdapter<GenericItem> {
+    return FastAdapter<GenericItem>().apply { block(this) }
+}
+
+fun <Item : GenericItem> fastadapter(block: (FastAdapter<Item>).() -> Unit): FastAdapter<Item> {
+    return FastAdapter<Item>().apply { block(this) }
+}
+
+fun <ParentItem : GenericItem, ChildItem : ParentItem> FastAdapter<ParentItem>.itemAdapter(block: ItemAdapter<ChildItem>.() -> Unit) {
```

Comment:
Given that you've put this in core, I think if you make this a direct method inside of `FastAdapter` instead of an extension function, you won't need the `ParentItem` generic. Up to you if that's something worth while

Reply:
![image](https://user-images.githubusercontent.com/1476232/59966574-5c344e00-951e-11e9-92b5-09d75217babb.png)


Doing this will result in a type being required. So the question is what is preferred :D

Reply:
You still need the `ChildItem` type, but you no longer need the `ParentItem` type right?

Reply:
exactly it would be like this inside the `FastAdapter`

```kotlin
    fun <ChildItem : Item> itemAdapter(block: ItemAdapter<ChildItem>.() -> Unit) {
        addAdapters(listOf(ItemAdapter<ChildItem>().apply(block) as IAdapter<Item>))
    }
```

so the question is what is preferred, the possibility of not passing in a type and using the one of the fastAdapter or the requirement to always specify

Reply:
I don't think it's useful information to always specify `ParentItem`, given that it can never deviate from the initial one provided right? At the same time, I do agree that DSL is best done as extensions where possible.

Not that this is any better, but this is another way:

```
fun <ParentItem : GenericItem> FastAdapter<ParentItem>.addAdapter(block: () -> IAdapter<out ParentItem>) {
    val adapter = block()
    addAdapters(listOf(adapter as IAdapter<ParentItem>))
}

fun <ChildItem : GenericItem> itemAdapter2(block: ItemAdapter<ChildItem>.() -> Unit): ItemAdapter<ChildItem> {
    return ItemAdapter<ChildItem>().apply(block)
}

fun t() {
    fastadapter<GenericItem> {
        addAdapter { 
            itemAdapter2<SimpleImageItem> { 
                
            }
        }
    }

}
```

I don't think you can avoid specifying the parent type if you have a single extension function that both adds the adapter and creates a new one

Reply:
I guess in this case the current approach we have is acceptable. 

In most cases the types will be specified from the FastAdapter anyways. 
In the more complex cases of multiple adapters with different types it should be acceptable to specify the parent and the child type